In [1]:
import pandas as pd
train_file = r"C:\Users\BHARGAVI\Downloads\project_data\ghc_train.csv"
test_file = r"C:\Users\BHARGAVI\Downloads\project_data\ghc_test.csv"

# Load CSV files into pandas DataFrames
train_df = pd.read_csv(train_file)
test_df=pd.read_csv(test_file)

# Preprocessing and Tokenizing for DL Models

In [2]:
import pandas as pd

# Function to map original labels to binary values
def map_labels(df):
    label_mapping = {'__label__0': 0, '__label__1': 1}
    if 'label' in df.columns:
        df['label'] = df['label'].map(label_mapping)
    return df

# Function to ensure text column is string type and handle missing values
def preprocess_text_column(df):
    df['text'] = df['text'].fillna('').astype(str)
    return df

# File paths
train_file = r"C:\Users\BHARGAVI\Downloads\project_data\ghc_train.csv"
test_file = r"C:\Users\BHARGAVI\Downloads\project_data\ghc_test.csv"

# Load CSV files into pandas DataFrames
train_df = pd.read_csv(train_file)
test_df = pd.read_csv(test_file)

# Apply functions to preprocess train_df and test_df
train_df = map_labels(train_df)
test_df = map_labels(test_df)

train_df = preprocess_text_column(train_df)
test_df = preprocess_text_column(test_df)


In [3]:
import pandas as pd

# Function to convert text column to lowercase
def convert_to_lower(df, text):
    df[text] = df[text].str.lower()
    return df

# File paths
train_file = r"C:\Users\BHARGAVI\Downloads\project_data\ghc_train.csv"
test_file = r"C:\Users\BHARGAVI\Downloads\project_data\ghc_test.csv"

# Load CSV files into pandas DataFrames
train_df = pd.read_csv(train_file)
test_df = pd.read_csv(test_file)

# Apply convert_to_lower function to 'text' column in train_df and test_df
train_df = convert_to_lower(train_df, 'text')
test_df = convert_to_lower(test_df, 'text')

# Example of using the modified train_df and test_df
print("Modified train_df:")
print(train_df.head())

print("\nModified test_df:")
print(test_df.head())


Modified train_df:
                                                text  hd  cv  vo  label
0  he most likely converted to islam due to his n...   0   0   0      0
1  so ford lied about being a psychologist. recor...   0   0   0      0
2     jobs. education. ending abuse of nation. ca43.   0   0   0      0
3  i share a lot of your values, & like many who ...   0   0   0      0
4  i am so ready to get back to blogging! www.ben...   0   0   0      0

Modified test_df:
                                                text  hd  cv  vo  label
0  https://www.youtube.com/watch?v=kacwpkaktak a ...   0   0   0      0
1  very nice! i tend to get tired of the constant...   0   0   0      0
2        watch today. https://circumcisionmovie.com/   0   0   0      0
3  " thinking venues " first color layer blocking...   0   0   0      0
4  what about death penalty for perpetrators  and...   0   0   0      0


In [4]:
import pandas as pd
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences

# Function to tokenize text data
def tokenize_texts(train_texts, test_texts, num_words=None):
    tokenizer = Tokenizer(num_words=num_words)
    tokenizer.fit_on_texts(train_texts)
    
    X_train_seq = tokenizer.texts_to_sequences(train_texts)
    X_test_seq = tokenizer.texts_to_sequences(test_texts)
    
    return tokenizer, X_train_seq, X_test_seq

# Function to pad sequences
def pad_sequences_data(X_train_seq, X_test_seq, max_sequence_length):
    X_train_padded = pad_sequences(X_train_seq, maxlen=max_sequence_length)
    X_test_padded = pad_sequences(X_test_seq, maxlen=max_sequence_length)
    return X_train_padded, X_test_padded

# Function to extract labels
def extract_labels(train_data, test_data):
    y_train = train_data['label'].values
    y_test = test_data['label'].values
    return y_train, y_test


In [5]:
# File paths
train_file = r"C:\Users\BHARGAVI\Downloads\project_data\ghc_train.csv"
test_file = r"C:\Users\BHARGAVI\Downloads\project_data\ghc_test.csv"

# Load CSV files into pandas DataFrames
train_df = pd.read_csv(train_file)
test_df = pd.read_csv(test_file)

# Example usage
tokenizer, X_train_seq, X_test_seq = tokenize_texts(train_df['text'], test_df['text'], num_words=5000)
X_train_padded, X_test_padded = pad_sequences_data(X_train_seq, X_test_seq, max_sequence_length=100)
y_train, y_test = extract_labels(train_df, test_df)

# Long Short-Term Memory (LSTM)

LSTM (Long Short-Term Memory) is a type of recurrent neural network (RNN) architecture designed to address the vanishing gradient problem that can occur when training traditional RNNs. It is particularly effective for modeling sequential data, where each element in the sequence depends on previous elements. 

In [6]:
import pandas as pd
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences
from keras.models import Sequential
from keras.layers import Embedding, LSTM, Dropout, Dense, SpatialDropout1D, Bidirectional
from sklearn.metrics import accuracy_score, classification_report

# Load datasets
train_df = pd.read_csv(r"C:\Users\BHARGAVI\Downloads\project_data\ghc_train.csv")
test_df = pd.read_csv(r"C:\Users\BHARGAVI\Downloads\project_data\ghc_test.csv")

# Tokenize texts
tokenizer = Tokenizer(num_words=5000)
tokenizer.fit_on_texts(train_df['text'])
X_train_seq = tokenizer.texts_to_sequences(train_df['text'])
X_test_seq = tokenizer.texts_to_sequences(test_df['text'])

# Pad sequences
X_train_padded = pad_sequences(X_train_seq, maxlen=100, padding='post')
X_test_padded = pad_sequences(X_test_seq, maxlen=100, padding='post')

# Extract labels
y_train = train_df['label'].values
y_test = test_df['label'].values

# Define LSTM model
lstm_model = Sequential()
lstm_model.add(Embedding(input_dim=5000, output_dim=128, input_length=100))
# Remove dropout after Embedding layer

# Add Bidirectional LSTM layer
lstm_model.add(Bidirectional(LSTM(100, dropout=0.2, recurrent_dropout=0.2)))
# Add Dropout layer after Bidirectional LSTM
lstm_model.add(Dropout(0.2))

lstm_model.add(Dense(1, activation='sigmoid'))  # For binary classification

# Compile the model
lstm_model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

# Train the model
lstm_model.fit(X_train_padded, y_train, epochs=5, batch_size=32, validation_data=(X_test_padded, y_test))

# Evaluate LSTM model
lstm_loss, lstm_accuracy = lstm_model.evaluate(X_test_padded, y_test)
lstm_y_pred = lstm_model.predict(X_test_padded)
lstm_y_pred_classes = (lstm_y_pred > 0.5).astype(int).flatten()  # Convert probabilities to binary classes
lstm_report = classification_report(y_test, lstm_y_pred_classes, target_names=['Class 0', 'Class 1'])

print("LSTM Model Accuracy:", lstm_accuracy)
print("LSTM Model Classification Report:\n", lstm_report)


C:\Users\BHARGAVI\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Epoch 1/5
689/689 ━━━━━━━━━━━━━━━━━━━━ 150s 208ms/step - accuracy: 0.8771 - loss: 0.3572 - val_accuracy: 0.8907 - val_loss: 0.2847
Epoch 2/5
689/689 ━━━━━━━━━━━━━━━━━━━━ 142s 207ms/step - accuracy: 0.9064 - loss: 0.2473 - val_accuracy: 0.8909 - val_loss: 0.2895
Epoch 3/5
689/689 ━━━━━━━━━━━━━━━━━━━━ 142s 206ms/step - accuracy: 0.9207 - loss: 0.2044 - val_accuracy: 0.8869 - val_loss: 0.3156
Epoch 4/5
689/689 ━━━━━━━━━━━━━━━━━━━━ 144s 210ms/step - accuracy: 0.9348 - loss: 0.1679 - val_accuracy: 0.8829 - val_loss: 0.3324
Epoch 5/5
689/689 ━━━━━━━━━━━━━━━━━━━━ 142s 206ms/step - accuracy: 0.9467 - loss: 0.1403 - val_accuracy: 0.8733 - val_loss: 0.3893
173/173 ━━━━━━━━━━━━━━━━━━━━ 8s 47ms/step - accuracy: 0.8695 - loss: 0.3929
173/173 ━━━━━━━━━━━━━━━━━━━━ 8s 43ms/step
LSTM Model Accuracy: 0.8733212351799011
LSTM Model Classification Report:
               precision    recall  f1-score   support

     Class 0       0.91      0.95      0.93      4803
     Class 1       0.51      0.38      0.43

# Convolutional Neural Network (CNN)

CNNs offer a powerful approach to text classification by leveraging their ability to learn hierarchical representations of sequential data. While traditionally associated with computer vision tasks, their adaptation to NLP tasks like text classification has proven effective, particularly when dealing with large datasets and complex patterns in text data.

In [8]:
import pandas as pd
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, Conv1D, MaxPooling1D, Flatten, Dense, Dropout
from sklearn.metrics import accuracy_score, classification_report

# Load datasets
train_df = pd.read_csv(r"C:\Users\BHARGAVI\Downloads\project_data\ghc_train.csv")
test_df = pd.read_csv(r"C:\Users\BHARGAVI\Downloads\project_data\ghc_test.csv")

# Tokenize texts
tokenizer = Tokenizer(num_words=5000)
tokenizer.fit_on_texts(train_df['text'])
X_train_seq = tokenizer.texts_to_sequences(train_df['text'])
X_test_seq = tokenizer.texts_to_sequences(test_df['text'])

# Pad sequences
X_train_padded = pad_sequences(X_train_seq, maxlen=100, padding='post')
X_test_padded = pad_sequences(X_test_seq, maxlen=100, padding='post')

# Extract labels
y_train = train_df['label'].values
y_test = test_df['label'].values

# Define CNN model
cnn_model = Sequential()
cnn_model.add(Embedding(input_dim=5000, output_dim=128, input_length=100))
cnn_model.add(Conv1D(128, 5, activation='relu'))
cnn_model.add(MaxPooling1D(5))
cnn_model.add(Conv1D(128, 5, activation='relu'))
cnn_model.add(MaxPooling1D(5))
cnn_model.add(Flatten())
cnn_model.add(Dense(128, activation='relu'))
cnn_model.add(Dropout(0.5))
cnn_model.add(Dense(1, activation='sigmoid'))  # For binary classification

# Compile the model
cnn_model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

# Train the model
cnn_model.fit(X_train_padded, y_train, epochs=5, batch_size=32, validation_data=(X_test_padded, y_test))

# Evaluate CNN model
cnn_loss, cnn_accuracy = cnn_model.evaluate(X_test_padded, y_test)
cnn_y_pred = cnn_model.predict(X_test_padded)
cnn_y_pred_classes = (cnn_y_pred > 0.5).astype(int).flatten()  # Convert probabilities to binary classes
cnn_report = classification_report(y_test, cnn_y_pred_classes, target_names=['Class 0', 'Class 1'])

print("CNN Model Accuracy:", cnn_accuracy)
print("CNN Model Classification Report:\n", cnn_report)



Epoch 1/5


C:\Users\BHARGAVI\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


689/689 ━━━━━━━━━━━━━━━━━━━━ 30s 40ms/step - accuracy: 0.8781 - loss: 0.3700 - val_accuracy: 0.8933 - val_loss: 0.2840
Epoch 2/5
689/689 ━━━━━━━━━━━━━━━━━━━━ 27s 39ms/step - accuracy: 0.9093 - loss: 0.2348 - val_accuracy: 0.8913 - val_loss: 0.2972
Epoch 3/5
689/689 ━━━━━━━━━━━━━━━━━━━━ 26s 38ms/step - accuracy: 0.9354 - loss: 0.1616 - val_accuracy: 0.8840 - val_loss: 0.4208
Epoch 4/5
689/689 ━━━━━━━━━━━━━━━━━━━━ 27s 39ms/step - accuracy: 0.9684 - loss: 0.0841 - val_accuracy: 0.8495 - val_loss: 0.5720
Epoch 5/5
689/689 ━━━━━━━━━━━━━━━━━━━━ 27s 39ms/step - accuracy: 0.9843 - loss: 0.0439 - val_accuracy: 0.8708 - val_loss: 0.9552
173/173 ━━━━━━━━━━━━━━━━━━━━ 2s 12ms/step - accuracy: 0.8690 - loss: 1.0048
173/173 ━━━━━━━━━━━━━━━━━━━━ 2s 11ms/step
CNN Model Accuracy: 0.8707804083824158
CNN Model Classification Report:
               precision    recall  f1-score   support

     Class 0       0.91      0.94      0.93      4803
     Class 1       0.50      0.37      0.42       707

    accura

# Gated Recurrent Unit (GRU)

Gated Recurrent Unit (GRU) is a type of recurrent neural network (RNN) architecture that addresses some of the shortcomings of traditional RNNs, such as the vanishing gradient problem.GRUs represent a powerful advancement in recurrent neural network architectures, offering improved performance and efficiency in sequence modeling tasks compared to traditional RNNs. Their ability to capture long-term dependencies while maintaining computational efficiency makes them a preferred choice for various NLP and time series prediction applications.








In [10]:
import pandas as pd
import numpy as np
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, GRU, Dense
from sklearn.metrics import classification_report

# Load datasets
train_df = pd.read_csv(r"C:\Users\BHARGAVI\Downloads\project_data\ghc_train.csv")
test_df = pd.read_csv(r"C:\Users\BHARGAVI\Downloads\project_data\ghc_test.csv")

# Tokenize texts
tokenizer = Tokenizer(num_words=5000)
tokenizer.fit_on_texts(train_df['text'])
X_train_seq = tokenizer.texts_to_sequences(train_df['text'])
X_test_seq = tokenizer.texts_to_sequences(test_df['text'])

# Pad sequences
max_sequence_length = 100
X_train_padded = pad_sequences(X_train_seq, maxlen=max_sequence_length, padding='post')
X_test_padded = pad_sequences(X_test_seq, maxlen=max_sequence_length, padding='post')

# Extract labels
y_train = train_df['label'].values
y_test = test_df['label'].values

# Define GRU model
gru_model = Sequential()
gru_model.add(Embedding(input_dim=5000, output_dim=128, input_length=max_sequence_length))
gru_model.add(GRU(units=128))  # GRU layer
gru_model.add(Dense(1, activation='sigmoid'))  # Output layer for binary classification

# Compile the model
gru_model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train the model
epochs = 5
batch_size = 32
gru_model.fit(X_train_padded, y_train, epochs=epochs, batch_size=batch_size, validation_data=(X_test_padded, y_test))

# Evaluate GRU model
gru_loss, gru_accuracy = gru_model.evaluate(X_test_padded, y_test)
gru_y_pred = gru_model.predict(X_test_padded)
gru_y_pred_classes = (gru_y_pred > 0.5).astype(int).flatten()  # Convert probabilities to binary classes
gru_report = classification_report(y_test, gru_y_pred_classes, target_names=['Class 0', 'Class 1'])

print("GRU Model Accuracy:", gru_accuracy)
print("GRU Model Classification Report:\n", gru_report)


Epoch 1/5


C:\Users\BHARGAVI\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


689/689 ━━━━━━━━━━━━━━━━━━━━ 81s 112ms/step - accuracy: 0.8722 - loss: 0.3825 - val_accuracy: 0.8717 - val_loss: 0.3836
Epoch 2/5
689/689 ━━━━━━━━━━━━━━━━━━━━ 77s 111ms/step - accuracy: 0.8799 - loss: 0.3685 - val_accuracy: 0.8717 - val_loss: 0.3858
Epoch 3/5
689/689 ━━━━━━━━━━━━━━━━━━━━ 78s 113ms/step - accuracy: 0.8783 - loss: 0.3721 - val_accuracy: 0.8717 - val_loss: 0.3847
Epoch 4/5
689/689 ━━━━━━━━━━━━━━━━━━━━ 77s 112ms/step - accuracy: 0.8776 - loss: 0.3632 - val_accuracy: 0.8900 - val_loss: 0.3029
Epoch 5/5
689/689 ━━━━━━━━━━━━━━━━━━━━ 79s 114ms/step - accuracy: 0.8973 - loss: 0.2623 - val_accuracy: 0.8924 - val_loss: 0.2889
173/173 ━━━━━━━━━━━━━━━━━━━━ 5s 27ms/step - accuracy: 0.8899 - loss: 0.2987
173/173 ━━━━━━━━━━━━━━━━━━━━ 5s 29ms/step
GRU Model Accuracy: 0.892377495765686
GRU Model Classification Report:
               precision    recall  f1-score   support

     Class 0       0.92      0.96      0.94      4803
     Class 1       0.63      0.40      0.49       707

    ac

The selected GRU (Gated Recurrent Unit) model has shown the highest accuracy among the compared models, achieving approximately 89.24%. It demonstrates good precision and recall metrics, particularly in distinguishing between the two classes (Class 0 and Class 1). The model's performance indicates its capability to effectively capture and utilize sequential dependencies in the data, making it well-suited for tasks such as sentiment analysis or sequence prediction. Compared to CNN and LSTM models, the GRU model performs better in terms of accuracy and metrics for Class 1, highlighting its strength in handling sequential data effectively.